### DISCUSSION TITLE
Automating PR checks using pre-commit hooks

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#281] [Ideas]
### Enforce coding stds, security, documentation guidelines etc for each Pull Request

We need to have some way of automating checks for pre-commit hooks. This ensures:

1. Formalised and non-negotiable minimum quality check for all Pull Requests
2. Implementation details clearly document the parameters which we are enforcing (i.e. you don't need a separate document which may be out of sync with the scripts). This will help developers to reduce frustration of rejected PRs.
3. Exponentially reduces maintainer effort in ensuring 'good' code. Maintainers can focus on logic while automation takes care of standards enforcement.
4. Eliminates manual errors resulting from fatigue etc. (Especially during deadlines or rush jobs)

https://www.npmjs.com/package/standard
https://hackernoon.com/enforcing-code-quality-for-node-js-c3b837d7ae17

There are quite a few tools. We need to evaluate and finalise for our project:
https://git-scm.com/docs/githooks
https://stackoverflow.com/questions/17947381/github-pull-request-hooks-static-code-analysis-and-provisional-rollback
https://githooks.com/
https://www.jetbrains.com/help/upsource/github-webhooks.html#set-up-a-project-specific-webhook-to-update-commits
Is this available for FOSS projects? : https://docs.github.com/en/enterprise-server@3.2/admin/policies/enforcing-policy-with-pre-receive-hooks/creating-a-pre-receive-hook-script

This is super informative and we need to go through this in a LOT of detail: https://github.com/aitemr/awesome-git-hooks

If we cannot configure hooks directly into github, we should write our own git-hooks which are automatically cloned to developer machines whenever they clone/fork etc.

**Comment by kcramakrishna** [comment]
Found some more:
https://www.npmjs.com/package/pre-commit
https://pre-commit.com/
https://metacpan.org/pod/App::GitHooks

(Unrelated python tool: https://pypi.org/project/hooks4git/)

**Comment by yureshwar** [comment]
Implemented eslint based configuration for source code analysis via git hooks. Followed the tutorial from the below given link

https://www.antstack.io/blog/adding-git-hooks-to-your-project/
